src/solvers/refinement/string_constraint_generator_main.cpp

```diff
@@ -513,54 +525,28 @@ exprt string_constraint_generatort::add_axioms_for_function_application(
 /// \par parameters: function application with one argument, which is a string,
 /// or three arguments: string, integer offset and count
 /// \return a new string expression
-string_exprt string_constraint_generatort::add_axioms_for_copy(
+exprt string_constraint_generatort::add_axioms_for_copy(
   const function_application_exprt &f)
 {
   const auto &args=f.arguments();
-  if(args.size()==1)
-  {
-    string_exprt s1=get_string_expr(args[0]);
-    return s1;
-  }
-  else
-  {
-    INVARIANT(
-      args.size()==3,
-      string_refinement_invariantt("f must have 1 or 3 arguments and the case "
-        "of 3 arguments is already handled"));
-    string_exprt s1=get_string_expr(args[0]);
-    exprt offset=args[1];
-    exprt count=args[2];
-    return add_axioms_for_substring(s1, offset, plus_exprt(offset, count));
-  }
+  PRECONDITION(args.size()==3 || args.size()==5);
+  const array_string_exprt res=char_array_of_pointer(args[1], args[0]);
+  const array_string_exprt str=get_string_expr(args[2]);
+  const typet &index_type=str.length().type();
+  const exprt offset=args.size()==3?from_integer(0, index_type):args[3];
+  const exprt count=args.size()==3?str.length():args[4];
+  return add_axioms_for_substring(res, str, offset, plus_exprt(offset, count));
 }
 
-/// for an expression of the form `array[0]` returns `array`
-/// \par parameters: an expression of type char
-/// \return an array expression
-exprt string_constraint_generatort::add_axioms_for_char_pointer(
-  const function_application_exprt &fun)
-{
-  exprt char_pointer=args(fun, 1)[0];
-  if(char_pointer.id()==ID_index)
-    return typecast_exprt(char_pointer.op0(), fun.type());
-  // TODO: It seems reasonable that the result of the function application
-  //       should match the return type of the function. However it is not
-  //       clear whether this typecast is properly handled in the string
-  //       refinement. We need regression tests that use that function.
-
-  // TODO: we do not know what to do in the other cases
-  TODO;
-  return exprt();
-}
 
 /// add axioms corresponding to the String.length java function
 /// \par parameters: function application with one string argument
 /// \return a string expression of index type
 exprt string_constraint_generatort::add_axioms_for_length(
```

Comment:
Is this function named appropriately? It doesn't add any axioms.

Reply:
You are right, the current implementation does not need to add axioms, it was named like that for uniformity with the other functions which converts function application.
Finding a better name for this would be a bit out of the scope of this PR.